### DISCUSSION TITLE
in what cases must fp64 be used to train the model

### DISCUSSION THREAD

**Discussion by ipe-zhangyz** [deepmodeling/deepmd-kit#1860] [Q&A]
It seems that the default precision of DeePMD-kit is double precision, or fp64 simply. However, in the SC20 Gordon Bell Prize paper the authors showed that for water system the test errors in energy and force are the same with fp32 and fp64. I want to know if there is any system that the fp64 is a must in terms of the accuracy of a model. Maybe it depends on the system that is studied and is hard to say, but I still wander if fp32 is accurate enough for most, for example 80% or 90% of cases?

**Comment by njzjz** [comment]
We benchmark FP64 and FP32 in [our new paper](https://arxiv.org/pdf/2304.09409.pdf). It looks like they give the same accuracy.

**Reply by ipe-zhangyz** [reply]
Thanks for your reply! In the BENCHMARKING paragraph of the new paper, it writes: We trained each model for a fixed number of steps (1 000 000 for water, Cu, and dipeptides, 16 000 000 for HEA, and 10 000 000 for OC2M and SPICE) **using neural networks** in double floating precision (FP64) and single floating precision (FP32) separately. Does it mean that only the neural networks, specifically speaking, the embedding net and fitting net, are tested using different floating precision? What about the the custom ops, as they also have implementions in different floating precisions, but using different precision setting method with the neural networks? In the SC20 paper https://ieeexplore.ieee.org/document/9355242, it is demonstrated clearly that when using Mixed-precision computation, atomic positions and the environment matrix construction use double precision in Sec. V-B3. If it is the same situation in the new paper, does it mean that the custom op must using fp64? Or is there any research on the effect to the final results of the simulation when using fp32 in custom ops? Thanks in advance!

**Reply by njzjz** [reply]
The paper only tests the precision of neural networks and its compressed form. The environment matrix is not tested.

**Reply by ipe-zhangyz** [reply]
So all the tests were conducted with environment matrix in fp64 precision, the same as the SC20 paper, right?

**Reply by njzjz** [reply]
The environment matrix is in FP64 precision. For the SC20 paper, you may contact @denghuilu.

**Reply by wanghan-iapcm** [reply]
In SC20, FP64 is used for the atomic positions and the environment matrix.

**Reply by ipe-zhangyz** [reply]
Thank you.  What I really concern is why you choose to use FP64 for the atomic positions and the environment matrix specifically, and not test FP32 as you do for the neural network parts. Could you please explain it a little more? Is there any theroy that makes the atomic positions and the environment matrix must use FP64, or you just use FP64 in accordance with practice in scientific computation?
